### PR TITLE
Use seek index to narrow index lookups

### LINKED ISSUE #3018: seek range in index objects
Search index objects should have an option to keep the seek range in the index.  This would speed up access a bit for needle searches.  When the seek range is large it can simply be omitted with a null value.

### PR DESCRIPTION
**Pull request by mattnibs**
Store the value count range for each key when creating indexes and use these use
 this range to narrow the scan of an index optimized query.

 This pr also plumbs the SeekIndexStride option into the branch.Load api
 which was necessary in order to get server tests working for this
 feature.

 Closes #3018

### CODE REVIEW

**Comment by mattnibs** [comment]
This pr is dependent on prs #3305 #3306. Since I can only set one branch as my base, the commit of #3306 is added to this pr but will be removed once #3306 is merged to main.

**Comment by mattnibs** [comment]
I'm going to move the strideIndex plumbing into a separate pr.

File: lake/index/expr.go [hunk 1]
```diff
@@ -112,15 +126,19 @@ func compareExpr(kv index.KeyValue, op string) (expr, error) {
...
 	for _, rule := range rules {
 		// XXX support indexes with multiple keys #3162
 		// and other rule types.
 		if fr, ok := rule.(*FieldRule); ok && in.Key.Equal(fr.Fields[0]) {
-			return rule
+			out := index.KeyValue{
+				Key:   append(field.New("key"), in.Key...),
+				Value: in.Value,
+			}
+			return out, rule
```
Comment (nwt) [inline lake/index/expr.go:138]: Unless you find this clearer with `out`, of course.
```suggestion
			return index.KeyValue{
				Key:   append(field.New("key"), in.Key...),
				Value: in.Value,
			}, rule
```

File: lake/index/filter.go [hunk 2]
```diff
@@ -2,15 +2,31 @@ package index
...
 )
 
+var (
+	minExpr = zedexpr.NewDotExpr(field.Dotted("seek.min"))
+	maxExpr = zedexpr.NewDotExpr(field.Dotted("seek.max"))
+	MaxSpan = extent.NewGenericFromOrder(
+		zed.NewUint64(0),
+		zed.NewUint64(math.MaxUint64),
+		order.Asc,
+	)
```
Comment (nwt) [inline lake/index/filter.go:27]: Nit: We don't usually wrap after each argument.
```suggestion
	MaxSpan = extent.NewGenericFromOrder(zed.NewUint64(0), zed.NewUint64(math.MaxUint64), order.Asc)
```

File: lake/index/filter.go [hunk 3]
```diff
@@ -31,27 +47,33 @@ func NewFilter(engine storage.Engine, path *storage.URI, dag *dag.Filter) (*Filt
...
 	u := ObjectPath(f.path, rid, oid)
 	finder, err := index.NewFinder(ctx, zed.NewContext(), f.engine, u)
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	rec, err := finder.Nearest(op, kv)
-	if err != nil {
-		return false, err
+	zv, err := finder.Nearest(op, kv)
```
Comment (nwt) [inline lake/index/filter.go:65]: Maybe change `zv` to `val` here and below? (We can always do that later, too.)

File: lake/seekindex/range.go [hunk 4]
```diff
@@ -17,6 +17,20 @@ func (r Range) Size() int64 {
 	return r.End - r.Start
 }
 
+func (r Range) Crop(r2 Range) Range {
```
Comment (nwt) [inline lake/seekindex/range.go:20]: Maybe put Crop next to Trim since they're similar?

File: lake/ztests/index/query-with-seek.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  zed create -q -seekstride 1KiB -orderby dname:asc test
+  zed use -q test@main
+  zed load -q  testscores.zson
+  zed query -s -z 'AvgScrMath == 306 | cut dname'
+  echo === | tee >(cat >&2)
+  zed index create -q AvgScrMath field AvgScrMath
+  zed index update -q AvgScrMath
+  zed query -s -z 'AvgScrMath == 306 | cut dname'
+
+
```
Comment (nwt) [inline lake/ztests/index/query-with-seek.yaml:13]:
```suggestion

```

File: service/ztests/index/query-with-seek.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  zed create -q -seekstride 1KiB -orderby dname:asc test
+  zed use -q test@main
+  zed load -q  testscores.zson
+  zed query -s -z 'AvgScrMath == 306 | cut dname'
+  echo === | tee >(cat >&2)
+  zed index create -q AvgScrMath field AvgScrMath
+  zed index update -q AvgScrMath
+  zed query -s -z 'AvgScrMath == 306 | cut dname'
+
+
```
Comment (nwt) [inline service/ztests/index/query-with-seek.yaml:12]:
```suggestion

```

**Review by nwt — APPROVED** [review]

**Pull request merged**